Dip/Dip/Dip.swift

```diff
@@ -161,37 +167,74 @@ public class DependencyContainer {
+    guard let definition = (self.definitions[key] ?? self.definitions[nilTagKey]) as? DefinitionOf<T> else {
+      fatalError("No definition registered with \(key) or \(nilTagKey)."
+        + "Check the tag, type you try to resolve, number, order and types of runtime arguments passed to `resolve()`.")
     }
-    return resolved
+
+    let usingKey: DefinitionKey? = definition.scope == .ObjectGraph ? key : nil
+    return _resolve(usingKey, definition: definition, builder: builder)
   }
   
   /// Actually resolve dependency
-  private func _resolve<T, F>(key: DefinitionKey, nilTagKey: DefinitionKey?, builder: F->T) -> T {
-    guard let definition = (self.dependencies[key] ?? self.dependencies[nilTagKey]) as? DefinitionOf<T> else {
-      fatalError("No instance factory registered with \(key) or \(nilTagKey)")
-    }
+  private func _resolve<T, F>(key: DefinitionKey?, definition: DefinitionOf<T>, builder: F->T) -> T {
     
-    if let resolvedInstance = definition.resolvedInstance {
-      return resolvedInstance
+    resolvedInstances.incrementDepth()
+    defer { resolvedInstances.decrementDepth() }
+    
+    if let previouslyResolved: T = resolvedInstances.previouslyResolved(key, definition: definition) {
+      return previouslyResolved
     }
     else {
-      let resolved = builder(definition.factory as! F)
-      definition.resolvedInstance = resolved
-      return resolved
+      let resolvedInstance = builder(definition.factory as! F)
+      
+      //when builder calls factory it will in turn resolve sub-dependencies (if there are any)
+      //when it returns instance that we try to resolve here can be already resolved
+      //so we return it, throwing away instance created by previous call to builder
+      if let previouslyResolved: T = resolvedInstances.previouslyResolved(key, definition: definition) {
+        return previouslyResolved
+      }
+      
+      resolvedInstances.storeResolvedInstance(resolvedInstance, forKey: key, definition: definition)
+      definition.resolveDependenciesBlock?(self, resolvedInstance)
+      
+      return resolvedInstance
     }
   }
   
   // MARK: - Private
   
-  private func lockAndDo(@noescape block: Void->Void) {
-    OSSpinLockLock(&lock)
-    defer { OSSpinLockUnlock(&lock) }
-    block()
+  let resolvedInstances = ResolvedInstances()
+  
+  ///Pool to hold instances, created during call to `resolve()`. 
+  ///Before `resolve()` returns pool is drained.
+  class ResolvedInstances {
+    var resolvedInstances = [DefinitionKey: Any]()
+    
+    func storeResolvedInstance<T>(instance: T, forKey key: DefinitionKey?, definition: DefinitionOf<T>) {
+      self.resolvedInstances[key] = instance
+      definition.resolvedInstance = instance
+    }
+    
+    func previouslyResolved<T>(key: DefinitionKey?, definition: DefinitionOf<T>) -> T? {
+      return (definition.resolvedInstance ?? self.resolvedInstances[key]) as? T
+    }
+    
+    var depth: Int = 0
+
+    func incrementDepth() {
```

Comment:
Actually without `defer` in `incrementDepth` it looks more straight forward to me =) Also `resolve` (or something like that) instead of `incrementDepth` looks like more descriptive for me, cause `incrementDepth` implies that there should be also decrement depth. Anyhow agree that this is better.